Initial checkin for MP Rest Client 3.0

andymc12 opened the pull request:
Initial implementation of epic issue #17947.

andymc12 committed:
Pull in RESTEasy MP Rest Client 4.7.0.Final modules

andymc12 committed:
Initial MP Rest Client 3.0 API (private) and public feature

dmuelle commented:
as far as I can tell, there are no new messages here that require ID review, just updating the existing for 3.0. If that's not correct, let me know.

Azquelt committed:
WIP: rest client 3.0

Azquelt committed:
Extend CDIExtensionMetadata to allow seeing app

Extend CDIExtensionMetadata to allow an extension to see beans from the
application (and from other extensions)

Azquelt committed:
More feature changes to get things working

Azquelt committed:
Fixes to get the TCK running

Reactive tests still do not run

Azquelt commented:
I've pushed additional changes, most tests should now pass with the exception of the reactive tests.

- Make it work without `restfulWS` being enabled in server.xml
  - refactor the resteasy CDI integration so that it can be used by both `restfulWS` and `mpRestClient`, but not `restfulWSClient`
- The CDI change to `CDIExtensionMetadata` would be an SPI change. We could use the old internal extension point if that's not desirable
  - The change is needed because resteasy mpRestClient beans expect to be able to see beans from the resteasy CDI integration, which is in a different bundle
- Get the reactive tests passing
  - As far as I can see, the reactive tests run on the server rather than the client, which is bad because that means it tests whatever implementation is on the test classpath, not whatever implementation is in the server.
  - We can probably get them passing by ensuring the right parts of resteasy are on the test classpath, but it would be better to change the TCK to run the tests on the server.

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/fat/src/org/eclipse/microprofile/rest/client/tck/RestClientTckPackageTest.java:
```diff
@@ -0,0 +1,71 @@
+        String javaVersion = System.getProperty("java.version");
+        Log.info(RestClientTckPackageTest.class, "setup", "javaVersion: " + javaVersion);
+        System.out.println("java.version = " + javaVersion);
+        if (javaVersion.startsWith("1.8")) {
```
We'll need to add another check here so that we only run the reactive TCK tests in FULL mode.

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/servers/FATServer/server.xml:
```diff
@@ -0,0 +1,28 @@
+        <feature>jsonb-2.0</feature>
+        <feature>mpConfig-3.0</feature>
+        <feature>mpRestClient-3.0</feature>
+        <feature>restfulWS-3.0</feature>
```
The plan is factor the RESTEasy CDI module out into its own bundle and include it in both the `restfulWS-3.0` feature and the `mpRestClient-3.0` feature.  At that point, we should not need to depend on `restfulWS-3.0` here.

andymc12 commented on dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.rest.client-3.0.feature:
```diff
@@ -0,0 +1,13 @@
+  io.openliberty.jakarta.annotation-2.0, \
+  io.openliberty.mpCompatible-5.0, \
+  io.openliberty.jakarta.cdi-3.0, \
+  io.openliberty.jakarta.restfulWS-3.0, \
```
This should be `restfulWSClient-3.0` once we refactor the RESTEasy CDI module into its own bundle.

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/pom.xml:
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2020 IBM Corporation and others. All rights reserved.
```
```suggestion
    Copyright (c) 2021 IBM Corporation and others. All rights reserved.
```

Azquelt replied:
2020 -> 2021?

You noted this before but it hasn't been changed.

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/pom.xml:
```diff
@@ -0,0 +1,72 @@
+
+    <properties>
+        <arquillian.version>1.7.0.Alpha10</arquillian.version>
+        <!-- arquillian.version>1.3.0.Final</arquillian.version -->
```
should remove this line

Azquelt replied:
Remove this line?

You noted this before but the change hasn't been made.

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/blank-tck-suite.xml:
```diff
@@ -0,0 +1,18 @@
+<!--Copyright (c) 2020 IBM Corporation and others. All rights reserved.
```
```suggestion
<!--Copyright (c) 2021 IBM Corporation and others. All rights reserved.
```

Can probably remove this file - but if we keep it, it should be accurate.

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/blank-tck-suite.xml:
```diff
@@ -0,0 +1,18 @@
+
+<suite name="microprofile-rest.client-1.2-tck" verbose="2"
+    configfailurepolicy="continue">
+    <test name="microprofile-rest.client-1.2-tck-all">
```
```suggestion
<suite name="microprofile-rest.client-3.0-tck" verbose="2"
    configfailurepolicy="continue">
    <test name="microprofile-rest.client-3.0-tck-all">
```

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+            </snapshots>
+        </repository>
+    </repositories>
+-->
```
remove?

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+    <properties>
+        <microprofile.rest.client.version>3.0-RC1</microprofile.rest.client.version>
+        <arquillian.version>1.7.0.Alpha10</arquillian.version>
+        <!-- arquillian.version>1.3.0.Final</arquillian.version -->
```
```suggestion
```

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+        </dependency>
+        -->
+        <!-- - - - - - - - - ????? - - - - - - - - - - - - -->
+
```
```suggestion
```

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+            <artifactId>javax.json</artifactId>
+            <version>1.1.2</version>
+        </dependency>
+        <!-- - - - - - - - - ????? - - - - - - - - - - - - -->
```
```suggestion
        <dependency>
            <groupId>org.glassfish</groupId>
            <artifactId>javax.json</artifactId>
            <version>1.1.2</version>
        </dependency>
```

tevans78 replied:
```suggestion
        <dependency>
            <groupId>org.glassfish</groupId>
            <artifactId>jakarta.json</artifactId>
            <version>2.0.1</version>
        </dependency>
```

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+            <version>1.2.0</version>
+            <scope>test</scope>
+        </dependency>
+         -->
```
```suggestion
```

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/src/test/java/com/ibm/ws/microprofile/rest/client/test/ArquillianLoadableExtension.java:
```diff
@@ -0,0 +1,24 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

Azquelt replied:
2020 -> 2021?

andymc12 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/tck-suite.xml:
```diff
@@ -0,0 +1,33 @@
+        </class>
+<!--         <class name="org.eclipse.microprofile.rest.client.tck.InvokeSimpleGetOperationTest"/> -->
+<!--         <class name="org.eclipse.microprofile.rest.client.tck.cditests.CDIInvokeSimpleGetOperationTest"/> -->
+<!--         <class name="org.eclipse.microprofile.rest.client.tck.cditests.HasSingletonScopeTest"/> -->
```
```suggestion
```

Azquelt replied:
Should these lines be removed?

andymc12 commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/io/openliberty/microprofile/rest/client30/internal/RestClientCdiExtensionMetadata.java:
```diff
@@ -0,0 +1,40 @@
+ * Registers the RestEasy MP Rest Client extension
+ */
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class RestClientCdiExtensionMetadata implements CDIExtensionMetadata {
```
Should this also make `RegisterRestClient` a bean defining annotation?

Azquelt replied:
If it does then we'll want a new test, given that it passes most of the TCKs without it.

Azquelt commented:
Remembered a thing: the jetty jars should be pulled in from maven rather than being checked in. May just be a case of adding them to the pom.xml?

andymc12 commented:
#build

tevans78 commented:
> Remembered a thing: the jetty jars should be pulled in from maven rather than being checked in. May just be a case of adding them to the pom.xml?

In addition to the 4 jetty jars, there are also two other jars which shouldn't be there...
```
dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/servers/FATServer/lib/global/httpcore-4.4.11.jar
dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/servers/FATServer/wiremock-standalone-2.14.0.jar
```

tevans78 commented on dev/com.ibm.ws.cdi.interfaces/src/io/openliberty/cdi/spi/CDIExtensionMetadata.java:
```diff
@@ -89,4 +89,11 @@
+    /**
+     * @return whether beans created from this extension can see and inject beans provided by the application and other extensions.
+     */
+    default boolean applicationBeansVisible() {
```
CDIExtensionMetadata is a public SPI and any changes need to be properly designed and reviewed via the UFO. I think that in this case it would be better to create a non-public extension interface and move this method to there. We then leave ourselves the option of making it public later.

Azquelt replied:
I've removed this change and added a separate `CDIExtensionMetadataInternal` interface.

tevans78 reviewed (changes requested):
CDIExtensionMetadata should not be changed. Use an internal interface instead.

tevans78 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/servers/FATServer/server.xml:
```diff
@@ -0,0 +1,28 @@
+-->
+<server>
+    <featureManager>
+        <feature>componenttest-1.0</feature>
```
```suggestion
        <feature>componenttest-2.0</feature>
```

tevans78 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+            <dependency>
+                <groupId>org.jboss.spec.javax.servlet</groupId>
+                <artifactId>jboss-servlet-api_3.0_spec</artifactId>
+                <version>1.0.2.Final</version> <!-- Needed for ZOS. Defines the version of this artifact that arquillian-liberty-managed should use -->
```
we don't want to use this because it is javax ... there is no jakarta equivalent so I don't know what to replace it with. In other TCKs we have just commented it out. Will wait and see if we get failures on z/OS.

tevans78 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/pom.xml:
```diff
@@ -0,0 +1,394 @@
+                        <!-- <tck_server>${tck_server}</tck_server> -->
+                        <tck_server>FATServer</tck_server>
+                        <!-- <tck_port>${tck_port}</tck_port>  -->
+                        <tck_port>8010</tck_port>
```
```suggestion
                        <tck_server>${tck_server}</tck_server>
                        <tck_port>${tck_port}</tck_port>
```

tevans78 commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/bnd.bnd:
```diff
@@ -0,0 +1,102 @@
+	com.ibm.websphere.javaee.servlet.4.0,\
+	com.ibm.websphere.javaee.validation.2.0;version=latest,\
+	io.openliberty.org.eclipse.microprofile.config.3.0;version=latest,\
+	io.openliberty.org.eclipse.microprofile.rest.client.3.0;version=latest,\
```
If this project is using MP 5.0 interfaces then it also needs to use JEE 9 interfaces; can't mix and match like this

andymc12 replied:
It should be using EE8 dependencies and then transforming.  I've changed it to use MP 4 instead of 5.

tevans78 commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/bnd.bnd:
```diff
@@ -0,0 +1,102 @@
+	com.ibm.ws.org.apache.httpcomponents,\
+	com.ibm.ws.org.glassfish.json.1.1
+
+jakartaeeMe: true
```
Are you sure that this bundle needs to be transformed?! I think that resteasy-client-microprofile already depends on jakarta interfaces?
https://repo1.maven.org/maven2/org/jboss/resteasy/resteasy-client-microprofile/4.7.0.Final/resteasy-client-microprofile-4.7.0.Final.pom

andymc12 replied:
Yes, per https://github.com/resteasy/resteasy/blob/cd4813a7260fd1f2f75fadcec055a430bdd6fcb4/resteasy-dependencies-bom/pom.xml#L34 - it is using CDI 2.0 (EE8).

tevans78 replied:
So the pom for the individual artifacts conflict with the overall bom?

Azquelt replied:
I don't think so. The individual artifact says "I need the Jakarta CDI API", and the BOM says "If you're using the Jakarta CDI API, you get version 2.0.2" (i.e. the one from Jakarta EE8)

skasund commented:
L2 message review not required.

andymc12 committed:
Update TCK dependencies to ensure we pass reactive TCK tests

- refactored RESTEasy CDI module to its own bundle
- pom.xml cleanup
- split reactive tests to only run in FULL mode

andymc12 commented:
Should pass TCK now - #build

andymc12 committed:
Code review comments and other cleanup

andymc12 commented:
Made additional changes for code review comments and other cleanup - better have #libby kick off another #build

andymc12 committed:
Download binaries rather than checking them in, and other cleanup

andymc12 commented:
Replaced checked-in binaries with downloaded artifacts. Updated package-info versions to match Andrew's changes in CDI. New #build

andymc12 commented:
Updated feature files to reference updated javadoc files (new file name due to new package/bundle version...).  New #build - #spawn.fullfat.buckets=io.openliberty.microprofile.rest.client.3.0.internal_fat_tck

andymc12 commented:
Previous build passed all tests, but it looks like the FeatureChecker task never completed - I don't see any logs or anything to indicate what went wrong...  will submit new #build to see if it was just some intermittent failure or not.

tevans78 commented on dev/cnf/oss_dependencies.maven:
```diff
@@ -356,10 +357,14 @@ org.eclipse.jetty.websocket:websocket-common:9.4.39.v20210325
 org.eclipse.jetty:jetty-util:9.2.2.v20140723
+org.eclipse.jetty:jetty-util:9.4.31.v20200723
 org.eclipse.jetty:jetty-util:9.4.39.v20210325
 org.eclipse.jetty:jetty-util:9.4.7.RC0
```
Please don't introduce additional minor versions when a later version is already available. There are already a bunch of issues/PRs open to update jetty due to CVEs. I suggest using version `9.4.39.v20210325` for now.

Azquelt replied:
@tevans78 These are test dependencies and this is the version that the TCK lists as its dependency. However, I don't think they need to be listed in this file as they're only referenced in the `build.gradle` file in the TCK project and not by any bnd files.

tevans78 commented on dev/cnf/oss_dependencies.maven:
```diff
@@ -280,6 +280,7 @@ org.apache.httpcomponents:httpclient:4.5.4
 org.apache.httpcomponents:httpcore-nio:4.4.13
 org.apache.httpcomponents:httpcore:4.1.2
+org.apache.httpcomponents:httpcore:4.4.11
 org.apache.httpcomponents:httpcore:4.4.13
```
please use the existing (higher) version

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/cdi-1.2/com.ibm.websphere.appserver.cdi-1.2.feature:
```diff
@@ -66,7 +66,7 @@ Subsystem-Name: Contexts and Dependency Injection 1.2
 -files=dev/api/ibm/schema/ibm-managed-bean-bnd_1_0.xsd, \
  dev/api/ibm/schema/ibm-managed-bean-bnd_1_1.xsd, \
- dev/spi/ibm/javadoc/com.ibm.websphere.appserver.spi.cdi_1.1-javadoc.zip
+ dev/spi/ibm/javadoc/com.ibm.websphere.appserver.spi.cdi_1.2-javadoc.zip
```
I don't want to change the SPI if we can avoid it

Azquelt replied:
I've removed the SPI change.

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/cdi-2.0/com.ibm.websphere.appserver.cdi-2.0.feature:
```diff
@@ -68,7 +68,7 @@ Subsystem-Name: Contexts and Dependency Injection 2.0
 -files=dev/api/ibm/schema/ibm-managed-bean-bnd_1_0.xsd, \
  dev/api/ibm/schema/ibm-managed-bean-bnd_1_1.xsd, \
- dev/spi/ibm/javadoc/com.ibm.websphere.appserver.spi.cdi_1.1-javadoc.zip
+ dev/spi/ibm/javadoc/com.ibm.websphere.appserver.spi.cdi_1.2-javadoc.zip
```
I don't want to change the SPI if we can avoid it

Azquelt replied:
I've removed the SPI change.

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/cdi-3.0/io.openliberty.cdi-3.0.feature:
```diff
@@ -68,7 +68,7 @@ Subsystem-Name: Jakarta Contexts and Dependency Injection 3.0
 -files=dev/api/ibm/schema/ibm-managed-bean-bnd_1_0.xsd, \
  dev/api/ibm/schema/ibm-managed-bean-bnd_1_1.xsd, \
- dev/spi/ibm/javadoc/com.ibm.websphere.appserver.spi.cdi_1.1-javadoc.zip
+ dev/spi/ibm/javadoc/com.ibm.websphere.appserver.spi.cdi_1.2-javadoc.zip
```
I don't want to change the SPI if we can avoid it ... although strictly this one should be the jakarta version anyway? @jhanders34 is that possible yet?

Azquelt replied:
I've removed the SPI change. The question of whether this can be the jakarta version of the javadoc instead still remains.

jhanders34 replied:
I am working on the API and SPI javadoc and pom file updates this week.  I have a pattern defined and just need to update the 11 or 12 api / spi projects to use it.  I can try to get one done and start a PR so you can see what I am thinking if you want.

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/mpRestClient-3.0/com.ibm.websphere.appserver.mpRestClient-3.0.feature:
```diff
@@ -0,0 +1,33 @@
+  io.openliberty.restfulWSClient-3.0, \
+  io.openliberty.mpConfig-3.0, \
+  io.openliberty.mpCompatible-5.0, \
+  com.ibm.websphere.appserver.org.eclipse.microprofile.rest.client-3.0, \
```
```suggestion
  io.openliberty.org.eclipse.microprofile.rest.client-3.0, \
```

tevans78 commented on dev/com.ibm.websphere.appserver.spi.cdi/bnd.bnd:
```diff
@@ -9,7 +9,7 @@
 #*******************************************************************************
 -include= ~../cnf/resources/bnd/bundle.props
-bVersion: 1.1
+bVersion: 1.2
```
I don't want to change the SPI if it can be avoided

Azquelt replied:
I've removed the SPI change.

tevans78 commented on dev/com.ibm.ws.cdi.interfaces/src/io/openliberty/cdi/spi/package-info.java:
```diff
@@ -11,9 +11,8 @@
  * @version 1.1
  */
-@org.osgi.annotation.versioning.Version("1.1")
+@org.osgi.annotation.versioning.Version("1.2")
```
I don't want to change the SPI if it can be avoided

Azquelt replied:
I've removed the SPI change.

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.rest.client-3.0.feature:
```diff
@@ -0,0 +1,13 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.org.eclipse.microprofile.rest.client-3.0
```
```suggestion
symbolicName=io.openliberty.org.eclipse.microprofile.rest.client-3.0
```

Azquelt commented:
@andymc12 @tevans78 I will work on removing change to the CDI SPI and provide an internal interface to use instead.

Azquelt committed:
Revert changes to CDI SPI package

Azquelt committed:
Add CDIExtensionMetadataInternal

New interface includes internal extensions to CDIExtensionMetadata.

Azquelt committed:
rest-client: use CDIExtensionMetadataInternal

The applicationBeansVisible method has been moved to the new
CDIExtensionMetadataInternal interface.

andymc12 committed:
Use io.openliberty for 3.0 features and remove test dependencies

andymc12 commented:
Addressed suggested changes of changing feature names to start with io.openliberty and removing test dependencies from oss_dependencies.maven. New #build

tevans78 commented on dev/io.openliberty.org.jboss.resteasy.cdi/bnd.bnd:
```diff
@@ -0,0 +1,103 @@
+# Using version=! in order to not have a version attached to the import for packages that were removed
+# from Java after Java 8.  Doing this keeps the import like before Java 11 support. It will get the
+# packages from Java when using Java 8 or earlier and from the new shipped bundles for Java 9 and later.
+#Import-Package: \
```
The # here looks wrong? Its either broken or the whole Import-Package header isn't needed?

tevans78 commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/org/jboss/resteasy/microprofile/client/RestClientBuilderImpl.java:
```diff
@@ -0,0 +1,785 @@
+package org.jboss.resteasy.microprofile.client;
```
Please add a copyright header even if the original didn't have one ... important to make it clear that this has a different copyright/license to the rest of OL. According to https://github.com/resteasy/resteasy/blob/3.0/License.html the header should be
```
  Copyright [yyyy] [name of copyright owner]

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
```

tevans78 replied:
Would also be useful to reference the upstream PR so that we know when the overlay can be deleted again.

tevans78 replied:
@andymc12 Sorry but please could you change `Copyright 2021 IBM Corporation and others.` to
`Copyright 2021 Red Hat, Inc., and individual contributors as indicated by the @author tags.` as they have it in other files e.g. https://github.com/resteasy/resteasy/blob/1cca43bb27e323873610754958f6f3ad46f0df9d/resteasy-core-spi/src/test/java/org/jboss/resteasy/spi/config/TestResteasyConfiguration.java#L1-L18

Also, the license says `You must cause any modified files to carry prominent notices stating that You changed the files` so underneath the copyright header please could you add a comment describing the changes you have made?

Azquelt replied:
Tom's follow up comments here still need to be addressed.

andymc12 committed:
Also update NLS filenames

andymc12 committed:
Remove unnecessary comments and add copyright header to overlay file

andymc12 commented:
New #build

andymc12 commented:
I believe all code review issues have been addressed. Please re-review.

Azquelt commented on dev/cnf/oss_dependencies.maven:
```diff
@@ -415,6 +415,7 @@ org.eclipse.microprofile.rest.client:microprofile-rest-client-api:1.2.0
 org.eclipse.microprofile.rest.client:microprofile-rest-client-api:1.3.1
 org.eclipse.microprofile.rest.client:microprofile-rest-client-api:1.4.0
 org.eclipse.microprofile.rest.client:microprofile-rest-client-api:2.0
+org.eclipse.microprofile.rest.client:microprofile-rest-client-api:3.0-m1
```
It looks like the latest release is RC3. Why are we using M1 rather than RC3?

andymc12 replied:
I don't anything has changed in the API since m1, but I don't mind switching to RC3.

Azquelt commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/fat/src/org/eclipse/microprofile/rest/client/tck/RestClientTckPackageTest.java:
```diff
@@ -0,0 +1,78 @@
+        server.startServer();
+    }
+
+    private static void useTCKSuite(String id) throws Exception {
+        Path cwd = Paths.get(".");
+        Log.info(RestClientTckPackageTest.class, "setup", "cwd = " + cwd.toAbsolutePath());
+        Path java8File = Paths.get("publish/tckRunner/tck/tck-suite.xml-" + id);
+        Path tckSuiteFile = Paths.get("publish/tckRunner/tck/tck-suite.xml");
+        Files.copy(java8File, tckSuiteFile, StandardCopyOption.REPLACE_EXISTING);
+    }
```
IMO it's slightly nicer to pass through a property to say which file to use rather than copying the files around but this does work.

Azquelt commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/servers/FATServer/.gitignore:
```diff
@@ -0,0 +1,6 @@
+jetty-io-9.4.31.v10100723.jar
+jetty-server-9.4.31.v10100723.jar
+jetty-util-9.4.31.v10100723.jar
+lib/global/*jar
```
Don't think this is needed any more since your gradle file only copies the libraries into the autoFVT directory.

andymc12 replied:
Correct.  Removing.

Azquelt commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/publish/tckRunner/tck/src/test/java/com/ibm/ws/microprofile/rest/client/test/ArquillianLoadableExtension.java:
```diff
@@ -0,0 +1,24 @@
+
+public class ArquillianLoadableExtension extends AbstractArquillianLoadableExtension {
+    @Override
+    public Set<TCKArchiveModifications> getModifications() {
+        return EnumSet.of(TCKArchiveModifications.JETTY, TCKArchiveModifications.TEST_LOGGER, TCKArchiveModifications.WIREMOCK);
+    }
```
I still think the TCK itself should include an extension which adds jetty and wiremock if they're required. Maybe worth looking at later?

andymc12 replied:
Sure

andymc12 committed:
More code review changes

Azquelt commented on dev/cnf/dependabot/check_this_in_if_it_changes/pom.xml:
```diff
@@ -2096,6 +2096,11 @@
+      <groupId>org.eclipse.microprofile.rest.client</groupId>
+      <artifactId>microprofile-rest-client-api</artifactId>
+      <version>3.0-m1</version>
+    </dependency>
```
Building locally should update this.

Emily-Jiang commented on dev/cnf/dependabot/check_this_in_if_it_changes/pom.xml:
```diff
@@ -2096,6 +2096,11 @@
+    <dependency>
+      <groupId>org.eclipse.microprofile.rest.client</groupId>
+      <artifactId>microprofile-rest-client-api</artifactId>
+      <version>3.0-m1</version>
```
we should use the version 3.0-RC3

Azquelt replied:
This has actually been changed in the oss_dependencies.maven file. A local build will update this file to match and then it needs committed.

Emily-Jiang commented on dev/com.ibm.ws.cdi.interfaces/src/com/ibm/ws/cdi/extension/CDIExtensionMetadataInternal.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * @return whether beans created from this extension can see and inject beans provided by the application and other extensions.
+     */
+    default boolean applicationBeansVisible() {
```
why not directly add this method to CDIExtensionMetadata?

Azquelt replied:
`CDIExtensionMetadata` is SPI.

I did originally add it there, but in an earlier review it was pointed out that that's an externals change and it would probably make sense to use it a bit internally first.

[See this comment](https://github.com/OpenLiberty/open-liberty/pull/18841#discussion_r727130532)

andymc12 commented:
#build

Azquelt reviewed (approved):
Looks good, great work.

andymc12 commented:
Single failure is unrelated to these changes. I'll plan to merge in an hour or two in case anybody else wants to re-review. #libby

jhanders34 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/bnd.bnd:
```diff
@@ -0,0 +1,29 @@
+#      or binaries from Artifactory (e.g. org.apache.derby:derbynet)
+# For all project names that match the pattern "*_fat*", dependencies for junit,
+# fattest.simplicity, and componenttest-1.0 will be automatically added to the buildpath
+-buildpath: \
```
This just seems wrong to have it end with `\` and then nothing.  Is it even needed?

andymc12 replied:
Works locally without it - so I removed it.

jhanders34 commented on dev/io.openliberty.microprofile.rest.client.3.0.internal_fat_tck/.settings/org.eclipse.jdt.ui.prefs:
```diff
@@ -0,0 +1,4 @@
+eclipse.preferences.version=1
+formatter_profile=_Andy'sProfile
```
Will this work on other people's envs?

andymc12 replied:
Dunno - I'm not a huge Eclipse fan - and definitely not a fan of checking in Eclipse-specific artifacts into GitHub - but it is what it is... Switching it back to Eclipse's default profile. Thanks for catching.

andymc12 committed:
Additional code review comments

andymc12 commented:
Will plan to merge tomorrow after next #build completes.  @tevans78 can you re-review (ideally today)?  GH won't let me merge until you approve. Thanks.

jhanders34 commented:
#libby
